README.md
logitr
================

[![Build
Status](https://travis-ci.org/MGallow/logitr.svg?branch=master)](https://travis-ci.org/MGallow/logitr)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/logitr)](https://cran.r-project.org/package=logitr)

### \*\*\*This package is experimental and not regularly maintained. It is recommended to use {glmnet} instead\!

## Overview

`logitr` is an R package for linear and logistic regression with
optional ridge and bridge regularization penalties. A (possibly
incomplete) list of functions contained in the package can be found
below:

  - `logisticr()` computes the coefficient estimates for logistic
    regression (ridge and bridge regularization optional)
  - `linearr()` computes the linear regression coefficient estimates
    (ridge regularization and weights optional)
  - `predict.logisticr()` generates predictions and loss metrics for
    logistic regression
  - `predict.linearr()` generates predictions and loss metric for linear
    regression

See [vignette](https://mgallow.github.io/logitr/) or
[manual](https://github.com/MGallow/logitr/blob/master/logitr.pdf).

## Installation

``` r
# The easiest way to install is from the development version from GitHub:
# install.packages("devtools")
devtools::install_github("MGallow/logitr")
```

If there are any issues/bugs, please let me know:
[github](https://github.com/MGallow/logitr/issues). You can also contact
me via my [website](https://mgallow.github.io/). Pull requests are
welcome\!

## Usage

``` r
library(logitr)

#we will use the iris data set
X = dplyr::select(iris, -c(Species, Sepal.Length))
y = dplyr::select(iris, Sepal.Length)
y_class = ifelse(dplyr::select(iris, Species) == "setosa", 1, 0)

#ridge regression (use CV for optimal lambda)
linearr(X, y, penalty = "ridge")
```

    ## 
    ## Call: linearr(X = X, y = y, penalty = "ridge")
    ## 
    ## Tuning parameters:
    ##       lam  alpha
    ## [1,]  0.1    NaN
    ## 
    ## MSE:
    ## [1] 0.09631325
    ## 
    ## Coefficients:
    ##               [,1]
    ## intercept  1.87785
    ##            0.64624
    ##            0.70231
    ##           -0.54160

``` r
#ridge logistic regression (IRLS) (use CV for optimal lambda)
logisticr(X, y_class, penalty = "ridge")
```

    ## 
    ## Call: logisticr(X = X, y = y_class, penalty = "ridge")
    ## 
    ## Iterations:
    ## [1] 18
    ## 
    ## Tuning parameters:
    ##       lam  alpha
    ## [1,]    0    NaN
    ## 
    ## MSE:
    ## [1] 8.0815e-14
    ## 
    ## logloss:
    ## [1] 6.929591e-06
    ## 
    ## misclassification:
    ## [1] 0
    ## 
    ## Coefficients:
    ##                [,1]
    ## intercept   7.88535
    ##             8.89485
    ##            -7.54863
    ##           -18.60425

``` r
#ridge logistic regression (MM)
logisticr(X, y_class, lam = 0.1, penalty = "ridge", method = "MM")
```

    ## 
    ## Call: logisticr(X = X, y = y_class, lam = 0.1, penalty = "ridge", method = "MM")
    ## 
    ## Iterations:
    ## [1] 5459
    ## 
    ## Tuning parameters:
    ##       lam  alpha
    ## [1,]  0.1    NaN
    ## 
    ## MSE:
    ## [1] 5.134801e-05
    ## 
    ## logloss:
    ## [1] 0.3956563
    ## 
    ## misclassification:
    ## [1] 0
    ## 
    ## Coefficients:
    ##               [,1]
    ## intercept  6.27623
    ##            1.54082
    ##           -3.64177
    ##           -1.63050

``` r
#bridge logistic regression (MM)
fit = logisticr(X, y_class, lam = 0.1, alpha = 1.2, penalty = "bridge")
```

    ## [1] "using MM algorithm..."

``` r
fit
```

    ## 
    ## Call: logisticr(X = X, y = y_class, lam = 0.1, alpha = 1.2, penalty = "bridge")
    ## 
    ## Iterations:
    ## [1] 26021
    ## 
    ## Tuning parameters:
    ##       lam  alpha
    ## [1,]  0.1    1.2
    ## 
    ## MSE:
    ## [1] 2.749702e-05
    ## 
    ## logloss:
    ## [1] 0.1878654
    ## 
    ## misclassification:
    ## [1] 0
    ## 
    ## Coefficients:
    ##               [,1]
    ## intercept 13.08031
    ##            0.61701
    ##           -5.71684
    ##           -0.23004

``` r
#predict using bridge logistic regression estimates
predict(fit, X[1:3,], y_class[1:3])
```

    ## $fitted.values
    ##           [,1]
    ## [1,] 0.9992467
    ## [2,] 0.9989747
    ## [3,] 0.9994881
    ## 
    ## $class
    ##      [,1]
    ## [1,]    1
    ## [2,]    1
    ## [3,]    1
    ## 
    ## $MSE
    ## [1] 6.269174e-07
    ## 
    ## $log.loss
    ## [1] 0.002291457
    ## 
    ## $misclassification
    ## [1] 0


NEWS.md
# logitr 0.1.0

* No major changes since logitr 0.1.0!





R/logistic.R
## Matt Galloway



#' @title Logistic Regression
#' @description Computes the coefficient estimates for logistic regression. ridge regularization and bridge regularization optional.
#'
#' @param X matrix or data frame
#' @param y matrix or vector of response values 0,1
#' @param lam optional tuning parameter(s) for ridge regularization term. If passing a list of values, the function will choose optimal value based on K-fold cross validation. Defaults to `lam = seq(0, 2, 0.1)`
#' @param alpha optional tuning parameter for bridge regularization term. If passing a list of values, the function will choose the optimal value based on K-fold cross validation. Defaults to 'alpha = 1.5'
#' @param penalty choose from c('none', 'ridge', 'bridge'). Defaults to 'none'
#' @param intercept Defaults to TRUE
#' @param method optimization algorithm. Choose from 'IRLS' or 'MM'. Defaults to 'IRLS'
#' @param tol tolerance - used to determine algorithm convergence. Defaults to 10^-5
#' @param maxit maximum iterations. Defaults to 10^5
#' @param vec optional vector to specify which coefficients will be penalized
#' @param init optional initialization for MM algorithm
#' @param criteria specify the criteria for cross validation. Choose from c('mse', 'logloss', 'misclass'). Defauls to 'logloss'
#' @param K specify number of folds for cross validation, if necessary
#' @return returns selected tuning parameters, beta estimates (includes intercept), MSE, log loss, misclassification rate, total iterations, and gradients.
#' @export
#' @examples
#' library(dplyr)
#' X = dplyr::select(iris, -Species)
#' y = dplyr::select(iris, Species)
#' y$Species = ifelse(y$Species == 'setosa', 1, 0)
#' logisticr(X, y)
#'
#' # ridge Logistic Regression with IRLS
#' logisticr(X, y, lam = 0.1, penalty = 'ridge')
#'
#' # ridge Logistic Regression with MM
#' logisticr(X, y, lam = 0.1, penalty = 'ridge', method = 'MM')

logisticr = function(X, y, lam = seq(0, 2, 0.1), alpha = 1.5, 
    penalty = c("none", "ridge", "bridge"), intercept = TRUE, 
    method = c("IRLS", "MM"), tol = 1e-05, maxit = 1e+05, 
    vec = NULL, init = 1, criteria = c("logloss", "mse", "misclass"), 
    K = 5) {
    
    # checks
    n = dim(X)[1]
    p = dim(X)[2]
    X = as.matrix(X)
    y = as.matrix(y)
    call = match.call()
    penalty = match.arg(penalty)
    method = match.arg(method)
    criteria = match.arg(criteria)
    lam = sort(lam)
    if (penalty == "none") {
        lam = 0
        alpha = 1.5
    }
    if (all(y == 1 | y == 0) == FALSE) 
        stop("y must be binary!")
    vec_ = vec
    if (is.null(vec)) {
        vec_ = rep(1, p)
    }
    if (all(alpha >= 2 | alpha <= 1)) 
        stop("alpha must be between 1 and 2!")
    if (all(lam >= 0) == FALSE) 
        stop("lam must be nonnegative!")
    if (all(lam == 0) & (penalty != "none")) {
        print("No penalty used: lam = 0")
        penalty = "none"
    }
    if ((penalty == "bridge") & (method != "MM")) {
        print("using MM algorithm...")
        method = "MM"
    }
    if (intercept) {
        # if no first column of ones, then add it
        if (all(X[, 1] != rep(1, n))) {
            X = cbind(1, X)
            p = dim(X)[2]
        }
        # do not penalize intercept, if not specified
        if (is.null(vec)) {
            vec_ = c(0, rep(1, p - 1))
        }
    }
    if (length(init) > 1) {
        if (p != length(init)) 
            stop("initialization wrong dimension!")
    }
    
    
    # CV needed?
    if ((length(lam) > 1 | length(alpha) > 1) & (penalty != 
        "none")) {
        
        # execute CV_logisticc
        CV = CV_logisticc(X, y, lam, alpha, penalty, intercept, 
            method, tol, maxit, vec_, init, criteria, K)
        lam = CV$best.lam
        alpha = CV$best.alpha
    }
    
    # execute logisticc
    logistic = logisticc(X, y, lam, alpha, penalty, intercept, 
        method, tol, maxit, vec_, init)
    
    
    # add intercept name, if needed
    betas = logistic$coefficients
    grads = logistic$gradient
    if (intercept == TRUE) {
        b1 = as.matrix(betas[1])
        rownames(b1) = "intercept"
        betas = rbind(b1, as.matrix(betas[-1, ]))
        g1 = as.matrix(grads[1])
        rownames(g1) = "intercept"
        grads = rbind(g1, as.matrix(grads[-1, ]))
    }
    
    # generate fitted values
    fit = predict_logisticc(logistic$coefficients, as.matrix(X), 
        y)
    
    # misc
    if (penalty == "none") {
        lam = NaN
    }
    if (penalty != "bridge") {
        alpha = NaN
    }
    parameters = matrix(c(lam, alpha), ncol = 2)
    colnames(parameters) = c("lam", "alpha")
    
    returns = list(call = call, parameters = parameters, coefficients = betas, 
        MSE = fit$MSE, log.loss = fit$log.loss, misclassification = fit$misclassification, 
        total.iterations = logistic$total.iterations, gradient = grads)
    class(returns) = "logisticr"
    return(returns)
    
}




##-----------------------------------------------------------------------------------




#' @title Print logisticr object
#' @description Print logisticr object.
#' @param x logisticr class object
#' @param ... additional arguments
#' @keywords internal
#' @export
print.logisticr = function(x, ...) {
    
    # print call
    cat("\nCall: ", paste(deparse(x$call), sep = "\n", collapse = "\n"), 
        "\n", sep = "")
    
    # print iterations
    cat("\nIterations:\n")
    print.default(x$total.iterations, quote = FALSE)
    
    # print optimal tuning parameters
    cat("\nTuning parameters:\n")
    print.default(round(x$parameters, 3), print.gap = 2L, 
        quote = FALSE)
    
    # print MSE
    cat("\nMSE:\n")
    print.default(x$MSE, quote = FALSE)
    
    # print logloss
    cat("\nlogloss:\n")
    print.default(x$log.loss, quote = FALSE)
    
    # print misclass
    cat("\nmisclassification:\n")
    print.default(x$misclassification, quote = FALSE)
    
    # print coefficients if dim <= 10
    if (nrow(x$coefficients) <= 20) {
        cat("\nCoefficients:\n")
        print.default(round(x$coefficients, 5))
    } else {
        cat("\n(...output suppressed due to large dimension!)\n")
    }
    
}
